include/structured/Dialect/Substrait/IR/SubstraitOps.td

```diff
@@ -0,0 +1,96 @@
+//===-- SubstraitOps.td - Substrait operations definitions -*- tablegen -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SUBSTRAIT_DIALECT_SUBSTRAIT_IR_SUBSTRAITOPS
+#define SUBSTRAIT_DIALECT_SUBSTRAIT_IR_SUBSTRAITOPS
+
+include "structured/Dialect/Substrait/IR/SubstraitDialect.td"
+include "structured/Dialect/Substrait/IR/SubstraitTypes.td"
+include "mlir/IR/CommonAttrConstraints.td"
+include "mlir/IR/OpBase.td"
+
+class Substrait_Op<string mnemonic, list<Trait> traits = []> :
+  Op<Substrait_Dialect, mnemonic, traits> {
+}
+
+//===----------------------------------------------------------------------===//
+// Constraints
+//===----------------------------------------------------------------------===//
+
+class RegionOf<Constraint condition> : Region<
+  Concat<"::llvm::all_of($_self.getOps(), [](::mlir::Operation &op) { "
+         "return ",
+         SubstLeaves<"$_self", "op", condition.predicate>,
+         "; })">,
+  "region where each child op is " # condition.summary>;
+
+class IsOp<string opType> : Constraint<
+  CPred<"::llvm::isa<" # opType # ">($_self)">,
+  "op of type '" # opType # "'">;
+
+class AnyOf<list<Constraint> opTypes, string summary = ""> : Constraint<
+  Or<!foreach(opType, opTypes, opType.predicate)>,
+     !if(!eq(summary, ""),
+        !interleave(!foreach(t, opTypes, t.summary), " or "),
+        summary)>;
+
+//===----------------------------------------------------------------------===//
+// Plan
+//===----------------------------------------------------------------------===//
+
+def PlanBodyOp : AnyOf<[
+    IsOp<"::mlir::substrait::PlanRelOp">
+  ]>;
+
+def Substrait_PlanOp : Substrait_Op<"plan", [
+    DeclareOpInterfaceMethods<OpAsmOpInterface, ["getDefaultDialect"]>,
+    NoTerminator, NoRegionArguments, SingleBlock
+  ]> {
+  let summary = "Represents a Substrait plan";
+  let description = [{
+    This op represents the `Plan` message type of Substrait. It carries the
+    version information inline as attributes, so it also subsumes the `Version`
+    message type. The body of the op consists of the `relation`s and (once
+    implemented) the extensions and types as well as their URLs defined in the
+    plan.
+  }];
+  let arguments = (ins
+    UI32Attr:$major_number,
+    UI32Attr:$minor_number,
+    UI32Attr:$patch_number,
+    DefaultValuedAttr<StrAttr, "\"\"">:$git_hash,
```

Comment:
What are these used for? (non blocking question)

Reply:
These come from the [`Version`](https://github.com/substrait-io/substrait/blob/main/proto/substrait/plan.proto#L60-L82) protobuf message type. The `git_hash` field is used to express in-between and forked versions; the `producer` field is used to document which tool produced the plan and allows to work around quirks of that producer.

I realized that the link between the dialect and the spec wasn't particularly explicit. I added a few sentences and links in the latest version. Let me know if that would have answered your question (or lead you to the right place).